Add internal `top_set_bit` function

LilithHafner opened the pull request:
At least 17 out of the 36 uses of `leading_zeros` in base are an attempt to compute how many bits are used. This PR defines that function explicitly and documents it for increased robustness and ease of use.

Commit:
add used_bits

Commit:
use used_bits

Commit:
document

Commit:
add tests

laborg commented:
I had need of such a function on occasions. With this being documented I assume its stable and can't be removed in future versions, right?

LilithHafner commented:
Yes. The documentation at https://docs.julialang.org is the authoritative source on what is guaranteed to be stable for all of 1.x (as described [here](https://docs.julialang.org/en/v1/manual/faq/#How-does-Julia-define-its-public-API?)). This PR would add `Base.used_bits` to that documentation.

Commit:
qualify name where needed (based on test failures)

oscardssmith commented:
I think we should probably triage this to bike-shop the name.

LilithHafner commented:
Triage is 1:15am my time, so I will not be joining you, but that seems like a good idea.

ararslan commented:
Seems like this could be exported given that `leading_*`, `trailing_*`, and `count_*` (for `*` in `zeros`, `ones`) are exported.

JeffBezanson commented:
Ideally this would be called `ndigits`; we should try to get that to constprop for constant base 2.

JeffBezanson commented:
If needed we can also call it `ndigits2` internally.

giordano commented:
`ndigits` has already fast-track for all bases which are powers of 2 (not just 2): #31722

MasonProtter commented:
The concern about `ndigits(x, base=2)` was that const-prop failed to inline one of the calls when this should really just compile down to a couple of assembly instructions, but it does still end up being quite fast even without the inlining:
```julia
julia> code_llvm(Tuple{Int}) do x
           ndigits(x, base=2)
       end
;  @ REPL[7]:2 within `#5`
define i64 @"julia_#5_386"(i64 signext %0) #0 {
top:
; ┌ @ intfuncs.jl:658 within `ndigits##kw`
; │┌ @ intfuncs.jl:658 within `#ndigits#426`
    %1 = call i64 @j_ndigits0z_388(i64 signext %0, i64 signext 2) #0
; ││┌ @ promotion.jl:488 within `max`
; │││┌ @ essentials.jl:489 within `ifelse`
      %2 = icmp sgt i64 %1, 1
      %3 = select i1 %2, i64 %1, i64 1
; └└└└
  ret i64 %3
}

julia> @benchmark ndigits(x, base=2) setup=(x=rand(1:10000))
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  4.740 ns … 8.200 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     5.390 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   5.137 ns ± 0.394 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▅                            █                             
  █▂▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▂█▄▂▁▁▁▁▁▁▁▂▁▁▁▁▁▁▁▂▁▂▂▂▂▁▁▂▂ ▂
  4.74 ns        Histogram: frequency by time       6.01 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```
I wonder if anyone is missing the `2ns` that could be gained here:
```julia
julia> used_bits(x::Base.BitInteger) = 8sizeof(x) - leading_zeros(x)
used_bits (generic function with 2 methods)

julia> code_llvm(Tuple{Int}) do x
           used_bits(x)
       end
;  @ REPL[13]:2 within `#9`
define i64 @"julia_#9_954"(i64 signext %0) #0 {
top:
; ┌ @ REPL[12]:1 within `used_bits`
; │┌ @ int.jl:421 within `leading_zeros`
    %1 = call i64 @llvm.ctlz.i64(i64 %0, i1 false)
; │└
; │┌ @ int.jl:86 within `-`
    %2 = sub nuw nsw i64 64, %1
; └└
  ret i64 %2
}

julia> @benchmark used_bits(x) setup=(x=rand(1:10000))
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  1.940 ns … 4.940 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.950 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.951 ns ± 0.062 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                                                          █  
  ▇▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▂▁▁▁▁█ ▂
  1.94 ns        Histogram: frequency by time       1.95 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

oscardssmith commented:
These 2ns matter a lot. Uses for this include things like `fmod` which in only a few ns anyway so this would be a roughly 100% regression.

giordano commented:
Would it be easier if for the time being the proposed `used_bits` function (maybe renamed to `ndigits2` as suggested above) wasn't part of the public API and kept only for internal usage when ultimate performance is needed?

KristofferC commented:
> Would it be easier if for the time being the proposed used_bits function (maybe renamed to ndigits2 as suggested above) wasn't part of the public API and kept only for internal usage when ultimate performance is needed?

Yes please. Always do things incrementally if possible imo.

Commit:
undocument

Commit:
rename to ndigits0z2

Commit:
update docstring

Commit:
update fallback to reflect documentation

oscardssmith commented:
I'm slightly worried that I will never remember that name.

LilithHafner commented:
Is inconsistency with `ndigits` when `x=0` something that triage discussed and decided was okay, or something that didn't come up? It seems misleading to call this `ndigits2` because of that inconsistency.

oscardssmith commented:
arg. Triage missed that. technically this corresponds to `ndigits(0,base=2,pad=0)`.

giordano commented on base/int.jl:
```diff
@@ -482,6 +482,30 @@ julia> trailing_ones(3)
+julia> ndigits0z2(-1)
+64
+"""
+ndigits0z2(x::Integer) = ndigits0z(x, 2)
```
Should these methods be moved to `intfuncs.jl` where all other `ndigits*` functions are defined?  This function isn't even used at all in this file, while `intfuncs.jl` is the main consumer (and tests are in `tests/intfuncs.jl`, too)

LilithHafner replied:
Yep! I placed that method here before I was aware of the `ndigits*` functions.

LilithHafner replied:
*Actually, this function is used by Core.Compiler and other consumers early in boostrapping so the `::BitInteger` method must stay in int.jl.

Commit:
move to intfuncs.jl

mcabbott commented on base/intfuncs.jl:
```diff
@@ -638,6 +638,31 @@ function ndigits0z(x::Integer, b::Integer)
+"""
+    ndigits0z2(x::Integer) -> Integer
+
+Efficiently compute `ndigits0z(x, 2)`. This is the number of bits used to represent `x` in
```
The first sentence isn't accurate, as they agree only for nonnegative numbers. 
```
julia> Base.ndigits0z(-10, 2)
4

julia> ndigits0z2(-10)
64
```
There is an example showing this, but best if the words aren't wrong.

LilithHafner replied:
Thanks! Good catch. This is actually worse than a documentation error because `ndigits0z2(big(-10)) == 4 != ndigits0z2(-10)`.

mcabbott replied:
I think that deleting the generic fallback means this function can't solve the original motivating problem:

https://github.com/JuliaLang/julia/pull/47519/files#diff-8a30f80e1bd3e2cedf6ca2fb06681ce8257bd55d7f92b75432e049f569321d1aR314-R316

I didn't notice earlier that the analogy with `Base.ndigits0z` had led to a different fallback. It doesn't matter for that use since there's an `abs`.

LilithHafner replied:
It can still help https://github.com/JuliaLang/julia/pull/47519/files#r1020836505.

You are right, though, that adding a generic fallback here would be able to replace the function defined in #47519 entirely. I think I'll wait for a name to come bac from triage before proceeding with the generic fallback.

LilithHafner replied:
I put back the generic fallback.

LilithHafner committed:
Fix error in docstring & remove generic fallback

The generic fallback is unused, untested, and was incorrect for negative numbers.

It's possible to get right, but I think a MethodError is fine.

If we revise this decision, I think `ceil(Integer, log2(x + oneunit(x)))` would be a correct general implementation.
And `x < 0 ? error() : x == 0 ? 0 : Int(ccall((:__gmpz_sizeinbase, :libgmp), Csize_t, (Base.GMP.MPZ.mpz_t, Cint), x, 2))` would be efficient for `BigInt`

LilithHafner commented:
This function differs from `ndigits` in a few ways
1) The base is always 2
2) `f(0) = 0` while `ndigits(0, 2) = 1`
3) `f(x < 0) = 8sizeof(x)` while `ndigits(x < 0, 2) = ndigits(abs(x), 2)` (thanks @mcabbott)

The name `ndigits2` is misleading for `x <= 0`

The name `ndigits0z2u` (👎) reflects all three differences using the suffix convention. "0z" - returns zero on input zero is already used in the ndigits* family, "2" - with a base of 2 is already used elsewhere (e.g. `log2`), "u" - unsigned. This name also demonstrates that the suffix convention should only be used in very limited cases, and multiple suffixes should not be combined.

In the context of sorting, the concept I was looking for is "how many bits do I need to radix over?"

Given that triage missed differences 2 (& 3?) when recommending the name `ndigits2`, I'm going to ask folks to reconsider.

mcabbott committed:
Compute `ceil(Int, log2(n))` in `TwicePrecision` using `leading_zeros` instead (#47519)

LilithHafner committed:
Merge branch 'master' into add-used-bits

Commit:
Replace, fix, & improve fallback; merge & fix tests

Commit:
replace ceil_log2 with ndigits0z2

Commit:
move the ::BitInteger method to int.jl for boostrapping and move ::BigInt to gmp.jl

LilithHafner committed:
Merge branch 'master' into add-used-bits

LilithHafner commented on test/ranges.jl:
```diff
@@ -1716,6 +1716,7 @@ end
     @test_throws BoundsError (false:true)[3]
 end
 
+
```
```suggestion
```

LilithHafner committed:
Update test/ranges.jl

Commit:
rename to top_set_bit, update documentation, and restrict to ::Integer

Commit:
documentation fixups

StefanKarpinski reviewed (approved):
This looks good to me and can be merged when green as far as I'm concerned. We might want to have a slightly broader name discussion before considering exporting, but I do think this would be a pretty hand function to export.

vchuravy commented on base/int.jl:
```diff
@@ -482,6 +482,33 @@ julia> trailing_ones(3)
+julia> top_set_bit(-1)
+64
+"""
+top_set_bit(x::BitInteger) = 8sizeof(x) - leading_zeros(x)
```
```suggestion
top_set_bit(x::BitInteger) = 8*sizeof(x) - leading_zeros(x)
```

LilithHafner commented on base/int.jl:
```diff
@@ -482,6 +482,33 @@ julia> trailing_ones(3)
+
+Negative `x` are only supported when `x::BitSigned`.
+
+`top_set_bit` is internal and will eventually be replaced by constant propagation.
+
```
```suggestion
```
The previous plan was to remove it once `Base.ndigits0z(x,2)` became equally as efficient due to constant propagation, but if folks think the end game may be exporting rather than removal (:-1: for increased code size) then maybe we should cut this line in the docstring.

oscardssmith replied:
I think cut the line.

LilithHafner replied:
`Base.ndigits0z(x,2)` and `top_set_bit(x)` do return the same result for all `x::BitUnsigned`

giordano commented:
Needs a rebase.

LilithHafner committed:
Update base/int.jl

LilithHafner committed:
Merge branch 'master' into add-used-bits
